lib/Twig/Extension/Core.php

```diff
@@ -821,10 +821,32 @@ function _twig_default_filter($value, $default = '')
 function twig_get_array_keys_filter($array)
 {
     if ($array instanceof Traversable) {
-        return array_keys(iterator_to_array($array));
+        if ($array instanceof IteratorAggregate) {
+            $array = $array->getIterator();
+        } // else $array must be an instance of IteratorAggregate here, we could call getIterator on that one etc. until we get an iterator
+
+        if ($array instanceof Iterator) {
+            $keys = [];
+            $array->rewind();
+            while($array->valid()) {
+                $keys[] = $array->key();
+                $array->next();
+            }
+
+            return $keys;
+        }
+
+        $keys = [];
```

Comment:
short array syntax cannot be used in Twig 1.x due to supported PHP versions (and should not be used in 2.x either to keep things maintainable when merging 1.x into 2.x)